### DISCUSSION TITLE
Fetch Interface Type from DataPortal

### DISCUSSION THREAD

**Discussion by coder2911** [MarimerLLC/csla#1818] [Questions]
Hi guys,

does anybody successfully fetch an interface type from DataPortal?

`var item = DataPortal.Fetch<IFake>(1);`

Inspired by the article [Abstractions in Csla](https://magenic.com/thinking/abstractions-in-csla) I did some testing using Csla version 5.3.1. 

I defined an IFake interface:

```c#
public interface IFake : IReadOnlyBase
{
    public int Id { get; }
    public string Name { get; }
}

```
and a fake root BO implementing my IFake interface :

```c#
[Serializable]
  public sealed class Fake : ReadOnlyBase<Fake>, IFake
  {
    public Fake() { }

    [Fetch]
    private void Fetch(int id)
    {
      Id = id;
      Name = "Fake";
    }

    public static readonly PropertyInfo<int> IdProperty = RegisterProperty<int>(nameof(Id));
    public int Id
    {
      get => GetProperty(IdProperty);
      private set => LoadProperty(IdProperty, value);
    }

    public static readonly PropertyInfo<string> NameProperty = RegisterProperty<string>(nameof(Name));
    public string Name
    {
      get { return GetProperty(NameProperty); }
      private set { LoadProperty(NameProperty, value); }
    }
  }

```
The TestActivator is implementing IDataPortalActivator to resolve the type and activate an instance of Fake BO:

```c#
internal sealed class TestActivator
    : IDataPortalActivator
  {
    public object CreateInstance(Type requestedType)
    {
        return Activator.CreateInstance(ResolveType(requestedType));
    }

    public void FinalizeInstance(object obj)
    {
      return;
    }

    public void InitializeInstance(object obj)
    {
      return;
    }

    public Type ResolveType(Type requestedType)
    {
      return typeof(Fake);
    }
  } 

```
 I created two simple tests:

```c#
[TestInitialize]
   public void Setup()
    {
       Csla.ApplicationContext.DataPortalActivator = new TestActivator();
    }

    [TestMethod]
    public void CanFetchType()
    {
      IFake item = Csla.DataPortal.Fetch<Fake>(1);
      Assert.IsNotNull(item);
    }

    [TestMethod]
    public void CanFetchInterface()
    {
      IFake item = Csla.DataPortal.Fetch<IFake>(1);
      Assert.IsNotNull(item);
    }
```
The first test  "CanFetchType" is successful and my TestActivator is working as expected, however the second test "CanFetchInterface" fails, when Server.DataPortal is trying to find the matching fetch method using the IFake type  in ServiceProviderMethodCaller.FindDataPortalMethod[T] resulting in a TargetParameterCountException. 

```text
Exception Message: 
    Test method Csla.Test.MyCsla.ActivatorTest.CanFetchInterface threw exception: 
    Csla.DataPortalException: DataPortal.Fetch failed (MyCsla.Test.IFake.[Fetch](Int32)) ---> System.Reflection.TargetParameterCountException: MyCsla.Test.IFake.[Fetch](Int32)
  Stack Trace: 
    ServiceProviderMethodCaller.FindDataPortalMethod[T](Type targetType, Object[] criteria, Boolean throwOnError) Line 230
    <Fetch>d__11.MoveNext() Line 268
     --- End of inner exception stack trace ---
    ServiceProviderMethodCaller.FindDataPortalMethod[T](Type targetType, Object[] criteria, Boolean throwOnError) Line 230
    <Fetch>d__11.MoveNext() Line 268
    DataPortal`1.Fetch(Type objectType, Object criteria) Line 413
    DataPortal`1.Fetch(Object[] criteria) Line 400
    DataPortal.Fetch[T](Object[] criteria) Line 221
    ActivatorTest.CanFetchInterface() Line 31
```

Is it possible to make DataPortal fetch an interface? 

Regards

**Comment by Chicagoan2016** [comment]
I hope @JasonBock  is reading this, it's a crying shame he is no longer with Magenic 😊

**Reply by rockfordlhotka** [reply]
He still checks in on #cslanet from time to time though, and he and I remain good friends 🙂

**Reply by JasonBock** [reply]
That article is pretty old at this point, and given how many changes CSLA has had in recent history, I wouldn't be surprised if things don't quite work with `IDataPortalActivator`. Sorry, can't really help here.

**Comment by rockfordlhotka** [comment]
It is possible that I broke something when rewriting the data portal for CSLA 5. The unit tests pass, but I'm not 100% sure there's a unit test for interfaces explicitly, though they are used [here](https://github.com/MarimerLLC/csla/blob/master/Source/Csla.test/Authorization/AuthTests.cs#L562) and [here](https://github.com/MarimerLLC/csla/blob/master/Source/Csla.test/DataPortal/InterceptorTests.cs#L236).

I wonder if the configuration change you make in the test init is actually taking effect? Can you put a breakpoint in your custom activator to make sure it is actually invoked?

**Reply by coder2911** [reply]
I put a breakpoint in custom activator. In the test "CanFetchType" the code in Csla.Server.SimpleDataPortal method
 
`public async Task<DataPortalResult> Fetch(Type objectType, object criteria, DataPortalContext context, bool isSync)` 

invokes the TestActivator in line 92
  
`obj = new DataPortalTarget(ApplicationContext.DataPortalActivator.CreateInstance(objectType));`

The breakpoint in TestActivator is hit and 

`return Activator.CreateInstance(ResolveType(requestedType));`

creates an instance of Fake as expected. 

Whereas in test "CanFetchInterface" the Server.DataPortal is never invoking the custom activator, it throws before reaching the relevant code and makes no use of the custom activator up to this point. When
  
`serviceProviderMethodInfo = Reflection.ServiceProviderMethodCaller.FindDataPortalMethod<FetchAttribute>(objectType, Server.DataPortal.GetCriteriaArray(criteria));` 

is called in Server.DataPortal in line 268 IFake is used as objectType to retrieve method information and fails.

**Reply by coder2911** [reply]
Thanks for the kind reply. I found a workaround by wrapping the Data Portal calls in an object factory as described in the article, but avoid to call the Data Portal methods with interface types. In the object factory I use reflection to call the Data Portal methods with the resolved type instead.
